src/state_machine/signer/mod.rs

```diff
@@ -260,20 +269,61 @@ impl<SignerType: SignerTrait> Signer<SignerType> {
             return Ok(Message::DkgEnd(DkgEnd {
                 dkg_id: self.dkg_id,
                 signer_id: self.signer_id,
-                status: DkgStatus::Failure("Bad state".to_string()),
+                status: DkgStatus::Failure(DkgFailure::BadState),
             }));
         }
 
         // only use the public shares from the DkgEndBegin signers
+        let mut missing_public_shares = HashSet::new();
+        let mut missing_private_shares = HashSet::new();
+        let mut bad_public_shares = HashSet::new();
         if let Some(dkg_end_begin) = &self.dkg_end_begin_msg {
             for signer_id in &dkg_end_begin.signer_ids {
-                let shares = &self.dkg_public_shares[signer_id];
-                for (party_id, comm) in shares.comms.iter() {
-                    self.commitments.insert(*party_id, comm.clone());
+                if let Some(shares) = self.dkg_public_shares.get(signer_id) {
+                    for (party_id, comm) in shares.comms.iter() {
+                        if !comm.verify() {
+                            bad_public_shares.insert(*signer_id);
+                        } else {
+                            self.commitments.insert(*party_id, comm.clone());
+                        }
+                    }
+                } else {
+                    missing_public_shares.insert(*signer_id);
+                }
+                if let Some(_shares) = self.dkg_private_shares.get(signer_id) {
+                    // TODO: consider move private shares checks here
+                } else {
+                    missing_private_shares.insert(*signer_id);
                 }
             }
         }
 
+        if !missing_public_shares.is_empty() {
```

Comment:
It seems like all of the above hashsets could be full simultaneously? In this case, I think it would be more important to report bad public shares in the error rather than the missing public vs private shares. Or does the order truly not matter/if we have MissingPublicShares, we will never have bad public shares?

Reply:
I did the checks in this order since we won't really know what shares are bad if we don't have them all.  And we need the public shares to check the private shares.  Hence public-missing -> public-bad -> private-missing -> private-bad ordering.

In practice we don't hit this function until `fn can_dkg_end` succeeds, so unless something is badly wrong the only things we will hit are bad public or private shares.

Reply:
Ah okay if they can't be known without having all shares, then sounds good. :D